### PR TITLE
[Seedless-onboarding] SMS OTP 

### LINKED ISSUE #2486: [Seedless-Onboarding] SMS OTP & Authenticator TOTP
## What is this feature about
This feature is about adding a recovery option via SMS OTP / Authenticator TOTP.

According to W3A there is a docker container which can be used for these recovery methods.

## Requirements
- [ ] SMS OTP
- [ ] Authenticator TOTP
- [ ] Design

### PR DESCRIPTION
**Pull request by schmanu**
## What it solves
Adds SMS OTP as auth method

Resolves #2486 

## How this PR fixes it
- New MFA recovery method SMS Recovery
- When recovering an Account the user has to choose from the available methods
- Integrate with Web3Auth SMS auth service

## Open Todos
- [x] Analytics
- [x] Tests
- [x] Manage factors in a reactive way (currently a refresh is necessary)
- [x] Encrypt factor key before storing it
- [x] Error handling

## How to test it
- Login with Google
- Navigate to Settings
- Setuo SMS OTP
- Delete local storage
- Disconnect
- Relogin and recover the account through SMS OTP

## Screenshots
Screenshot 2023-11-23 at 22 17 44
Screenshot 2023-11-23 at 22 17 50
Screenshot 2023-11-23 at 22 19 54
Screenshot 2023-11-23 at 22 20 00

## Checklist
* [x] I've tested the branch on mobile 📱
* [x] I've documented how it affects the analytics (if at all) 📊
* [x] I've written a unit/e2e test for it (if applicable) 🧑‍💻

### CODE REVIEW

File: src/components/settings/SecurityLogin/SmsMfaForm/index.tsx [hunk 1]
```diff
@@ -0,0 +1,211 @@
...
+                        readOnly: verificationStarted || Boolean(currentNumber),
+                      }}
+                      {...register(SmsOtpFieldNames.mobileNumber, {
+                        required: true,
+                      })}
+                    />
+                  </FormControl>
+                  {verificationStarted && (
+                    <Box>
+                      <Typography variant="h4">Verification Code</Typography>
```
Comment (katspaugh) [inline src/components/settings/SecurityLogin/SmsMfaForm/index.tsx:133]:
```suggestion
                      <Typography variant="h4" mb={1}>Verification code</Typography>
```

File: src/components/settings/SecurityLogin/SmsMfaForm/index.tsx [hunk 2]
```diff
@@ -0,0 +1,211 @@
...
+  return (
+    <FormProvider {...formMethods}>
+      <form onSubmit={handleSubmit(onSubmit)}>
+        <Box display="flex" flexDirection="column" gap={3} alignItems="baseline">
+          <Typography>Protect your social login signer with mobile recovery</Typography>
+          <Accordion expanded={open} defaultExpanded={false} onChange={toggleAccordion}>
+            <AccordionSummary expandIcon={<ExpandMoreIcon />}>
+              <Box display="flex" alignItems="center" gap={1}>
+                <SvgIcon component={CheckIcon} sx={{ color: isSmsOtpSet ? 'success.main' : 'border.light' }} />
+                <Typography fontWeight="bold">SMS Recovery</Typography>
```
Comment (katspaugh) [inline src/components/settings/SecurityLogin/SmsMfaForm/index.tsx:112]:
```suggestion
                <Typography fontWeight="bold">SMS recovery</Typography>
```

File: src/services/mpc/PasswordRecoveryModal.tsx [hunk 3]
```diff
@@ -13,30 +22,146 @@ export const close = () => {
...
+              Choose your recovery method
+            </Typography>
+            <Typography>
+              This browser is not registered with your Account yet. Please proceed with one of your setup recovery
+              methods.
+            </Typography>
+          </Box>
+          <Divider />
+          <Box p={4}>
+            <Typography>Use your mobile phone</Typography>
```
Comment (katspaugh) [inline src/services/mpc/PasswordRecoveryModal.tsx:54]:
```suggestion
            <Typography mb={1}>Use your mobile phone</Typography>
```

File: src/services/mpc/PasswordRecoveryModal.tsx [hunk 4]
```diff
@@ -13,30 +22,146 @@ export const close = () => {
...
+            <Button onClick={() => setRecoveryMethod(RecoveryMethod.SMS)} disabled={!smsEnabled} variant="contained">
+              SMS
+            </Button>
+
+            <Typography mt={3}>Use your recovery password</Typography>
```
Comment (katspaugh) [inline src/services/mpc/PasswordRecoveryModal.tsx:59]:
```suggestion
            <Typography mt={3} mb={1}>Use your recovery password</Typography>
```

File: src/services/mpc/PasswordRecoveryModal.tsx [hunk 5]
```diff
@@ -13,30 +22,146 @@ export const close = () => {
...
+              onClick={() => setRecoveryMethod(RecoveryMethod.PASSWORD)}
+              disabled={!passwordEnabled}
+              variant="contained"
+            >
+              Password
+            </Button>
+
+            {!IS_PRODUCTION && (
+              <>
+                <Typography mt={3}>Only in Dev / Staging</Typography>
```
Comment (katspaugh) [inline src/services/mpc/PasswordRecoveryModal.tsx:70]:
```suggestion
                <Typography mt={3} mb={1}>Only in Dev / Staging</Typography>
```

File: src/services/mpc/recovery/SmsOtpRecovery.ts [hunk 6]
```diff
@@ -0,0 +1,292 @@
...
+    if (!storedFactorKey) {
+      currentStorage.remove('sms_tracking_id')
+      return false
+    }
+
+    const newFactorKey = new BN(storedFactorKey, 'hex')
+    if (isFirstTime) {
+      const storeData: SmsOtpStoreData = {
+        number,
+        moduleName: SMS_OTP_MODULE_NAME,
```
Comment (arch1995) [inline src/services/mpc/recovery/SmsOtpRecovery.ts:277]: Pls use `module` as key here.

File: src/services/mpc/recovery/SmsOtpRecovery.ts [hunk 7]
```diff
@@ -0,0 +1,292 @@
...
+  code: string
+
+  // custom metadata you want to set with the user.
+  // this is only for the first time, when user is setting up the MFA for the first time.
+  // Send the factorKey in this, which could be in raw format or encrypted using the postbox key.
+  // This can always be updated in the future as well.
+  data?: string
+}
+
+export const SMS_OTP_MODULE_NAME = 'SmsRecovery'
```
Comment (arch1995) [inline src/services/mpc/recovery/SmsOtpRecovery.ts:68]: Pls change this value to `external_sms`, this is to keep it in sync with the PnP version

File: src/services/mpc/recovery/SmsOtpRecovery.ts (hunk 6 above)
Comment (schmanu) [reply src/services/mpc/recovery/SmsOtpRecovery.ts]: Done. Thanks for the feedback :)

**Comment by schmanu** [comment]
Closed this PR in favor of #2968 
Branch had to be renamed to make the deployment work.

**Pull request closed without merging**